langchain4j-vertex-ai-gemini/src/main/java/dev/langchain4j/model/vertexai/VertexAiGeminiChatModel.java

```diff
@@ -101,16 +99,51 @@ public VertexAiGeminiChatModel(GenerativeModel generativeModel,
 
     @Override
     public Response<AiMessage> generate(List<ChatMessage> messages) {
-
         List<Content> contents = ContentsMapper.map(messages);
-
         GenerateContentResponse response = withRetry(() -> generativeModel.generateContent(contents, generationConfig), maxRetries);
 
-        return Response.from(
+        Content content = ResponseHandler.getContent(response);
+        Optional<Part> containsFunctionCall = content.getPartsList().stream().filter(Part::hasFunctionCall).findFirst();
+
+        if (containsFunctionCall.isPresent()) {
+            FunctionCall functionCall = containsFunctionCall.get().getFunctionCall();
+            ToolExecutionRequest toolExecutionRequest = FunctionCallHelper.fromFunctionCall(functionCall);
+
+            return Response.from(
+                AiMessage.from(toolExecutionRequest),
+                TokenUsageMapper.map(response.getUsageMetadata()),
+                FinishReasonMapper.map(ResponseHandler.getFinishReason(response))
+            );
+        } else {
+            return Response.from(
                 AiMessage.from(ResponseHandler.getText(response)),
                 TokenUsageMapper.map(response.getUsageMetadata()),
                 FinishReasonMapper.map(ResponseHandler.getFinishReason(response))
-        );
+            );
+        }
+    }
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, List<ToolSpecification> toolSpecifications) {
+        try {
+            Tool tool = FunctionCallHelper.convertToolSpecifications(toolSpecifications);
+            //TODO: hack for now, till the next release that will have a tool param
+            // in the underlying generateContent() method
+            this.generativeModel.setTools(Collections.singletonList(tool));
+            return generate(messages);
+        } finally {
+            //TODO: remove the finally block (see above todo)
+            this.generativeModel.setTools(Collections.emptyList());
+        }
+    }
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages, ToolSpecification toolSpecification) {
+        if (toolSpecification == null) {
+            return generate(messages);
+        } else {
+            return generate(messages, Collections.singletonList(toolSpecification));
```

Comment:
Does Gemini has a way to _force_ the model to call the provided tool?

The idea of this method (`generate(List<ChatMessage>, ToolSpecification)`) was to guarantee that provided `ToolSpecification` will be called.

Reply:
Currently, there's no way to enforce a tool to be called, but maybe it'll come later.